### DISCUSSION TITLE
gradient clipping

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2724] [Q&A] [answered]
Hi,
This is how I clip the gradients, but it seems that it will not work.
```rust
let loss = MseLoss::new().forward(output.clone(), batch.targets, Mean);
let grads = loss.backward();
let grads = GradientClippingConfig::Norm(1.0).init().clip_gradient(grads);
```
So could you give an example on how to clip the gradients ?

**Comment by wangjiawen2013** [comment]
Besides, how to display and get the values of the grads ?

**Reply by laggui** [reply, marked as answer]
With the `GradientsParams`, you can retrieve the gradients for a given `ParamId` (associated to a parameter tensor) with `grads.get(id)`.

And then you can display / check the values of the grad tensor. If you want to check all values of a module, you could use a `ModuleVisitor` and do something similar to the optimizer adaptor I linked in my other response.

**Comment by laggui** [comment]
You can set this via the optimizer config [`.with_gradient_clipping`](https://burn.dev/docs/burn/optim/struct.SgdConfig.html#method.with_gradient_clipping).

You can see how it's applied in the optimizer:
https://github.com/tracel-ai/burn/blob/b33bd24f88db3bf59519bb17378515ed829da05f/crates/burn-core/src/optim/simple/adaptor.rs#L120-L131

**Reply by wangjiawen2013** [reply]
Thanks, I got access to the grads and clipped them using the following code:
```rust
let tensor = Tensor::<Autodiff<NdArray>, 2>::from_floats([[1,2,3], [4,5,6]], &device);
let mut model: Linear<Autodiff<NdArray>> = LinearConfig::new(3, 1).init(&device);
let output = model.forward(tensor);
let loss = MseLoss::new().forward(output.clone(), Tensor::<Autodiff<NdArray>,2>::from_floats([[4], [7]], &device), Mean);
let grads = loss.backward();
let grads = GradientsParams::from_grads(grads, &model);

// Get access to the grads, NdArray must be used here.
println!("{}", grads.get::<NdArray, 2>(model.weight.id).unwrap());

let mut optim: OptimizerAdaptor<Adam, Linear<Autodiff<NdArray>>, Autodiff<NdArray>> = AdamConfig::new()
    .with_grad_clipping(Some(GradientClippingConfig::Norm(1.0)))
    .init();
model = optim.step(0.001, model, grads);
```
I can run the code, and I found we must use `NdArray` when getting the grads, `Autodiff<NdArray>` will cause an panic.
